### PR TITLE
🎉 Integration Testing for SSH using a docker container | Postgres Source and Destination update integration tests using ssh bastion in docker container

### LINKED ISSUE #5811: Integration Testing for SSH using a docker container
## Tell us about the problem you're trying to solve
Improve the process of acceptance testing by allowing doing acceptance / integration testing using a bastion running in a docker container. Right now we have a dedicated host in AWS that we use. 

Why do we care?
1. We'd rather not maintain the bastion infra if we don't have to 
2. It means that the tests require secrets as an input which adds a bit more friction. If we use docker that would not be necessary.
3. It requires having a static instance of the database we trying to test running! Because the bastion is static it also needs to be forwarding to some database in our VPC so we need a statically running db as well. This limits the tests we can right and means we're paying for and maintaining a database for no good reason. Ideally we could spin up dbs and bastion all in docker!

@sherifnada fyi

### PR DESCRIPTION
**Pull request by VitaliiMaltsev**
## What
Improve the process of acceptance testing by allowing doing acceptance / integration testing using a bastion running in a docker container. Right now we have a dedicated host in AWS that we use.

Why do we care?

1. We'd rather not maintain the bastion infra if we don't have to

2. It means that the tests require secrets as an input which adds a bit more friction. If we use docker that would not be necessary.
3. It requires having a static instance of the database we trying to test running! Because the bastion is static it also needs to be forwarding to some database in our VPC so we need a statically running db as well. This limits the tests we can right and means we're paying for and maintaining a database for no good reason. Ideally we could spin up dbs and bastion all in docker!

## How
A new SshBastion class has been implemented, which allows you to initialize and run a bastion ssh server created in a docker test container. The test database and bastion run in containers on the same network. Connection configuration for integration tests is now taken directly from container settings and does not require a real database connection. Establishing an ssh tunnel implies 2 authentication options - using a password or using a private key

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by alexandr-shegeda — APPROVED** [review]
LGTM

**Review by DoNotPanicUA — APPROVED** [review]

**Comment by DoNotPanicUA** [comment]
I would only recommend not to increase the version at files `airbyte-config/init/src/main/resources/seed/destination_definitions.yaml` like that before you get external approval. If you increase the version before pulling the image, builds, and tests fail.

File: docs/integrations/sources/postgres.md
Comment (sherifnada) [inline docs/integrations/sources/postgres.md:267]: this should go in the README for the connector rather than this doc. This doc is meant for consumption by Airbyte users, and this is an implementation detail of how we test the connector

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java
Comment (sherifnada) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java:45]: why is this all static instead of instance variables?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java
Comment (sherifnada) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java:40]: I suggest adding the word `Container` to the. name e.g: `SshBastionContainer` to make it clear it's containerized

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractSshPostgresSourceAcceptanceTest.java
Comment (sherifnada) [inline airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractSshPostgresSourceAcceptanceTest.java:27]: style: prefer using explicit import instead of star imports

**Review by sherifnada — COMMENTED** [review]
Looks great! a few comments: 

1. Is there a reason we can't treat the bastion as a normal object the same way other test containers are? Right now it's static/global which could introduce isolation issues
2. could you remove the places where we inject secrets in `ci_credentials.sh` for the current postgres ssh test case and once you merge this pr remove them from environment secrets as well in github? 
3. you actually don't need to bump the postgres container versions since this is just changing their tests not the connector itself

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractSshPostgresSourceAcceptanceTest.java
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractSshPostgresSourceAcceptanceTest.java]: > style: prefer using explicit import instead of star imports

replaced star import with explicit

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java
Comment (VitaliiMaltsev) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java]: > I suggest adding the word `Container` to the. name e.g: `SshBastionContainer` to make it clear it's containerized

good point! Renamed class to SshBastionContainer

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java
Comment (VitaliiMaltsev) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/SshBastion.java]: > why is this all static instead of instance variables?

replaced with instance variables

File: docs/integrations/sources/postgres.md
Comment (VitaliiMaltsev) [reply docs/integrations/sources/postgres.md]: > this should go in the README for the connector rather than this doc. This doc is meant for consumption by Airbyte users, and this is an implementation detail of how we test the connector

removed description from this doc

**Comment by VitaliiMaltsev** [comment]
> Looks great! a few comments:
> 
> 1. Is there a reason we can't treat the bastion as a normal object the same way other test containers are? Right now it's static/global which could introduce isolation issues
> 2. could you remove the places where we inject secrets in `ci_credentials.sh` for the current postgres ssh test case and once you merge this pr remove them from environment secrets as well in github?
> 3. you actually don't need to bump the postgres container versions since this is just changing their tests not the connector itself

1. Yes, you are right. Replaced static initialization with initialization of the new instance of class
2. Done
3. Revert changes with version bumping

**Review by sherifnada — APPROVED** [review]
Great work!

**Comment by cgardens** [comment]
@VitaliiMaltsev this is awesome! great work.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/readme.md
Comment (ChristopheDuong) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/ssh/readme.md:59]: - Does this mean we can update the `Future Work` section and remove the bullet point?
> Improve the process of acceptance testing by allowing doing acceptance testing using a bastion running in a docker container instead of having to use dedicated infrastructure and a static database.

- How could we enable these ssh integration tests to other destinations too?
- Would these integration tests catch errors in ssh tunnels handling with normalization too? see https://github.com/airbytehq/oncall/issues/145

**Pull request merged**
